distributed/comm/core.py

```diff
@@ -274,68 +274,59 @@ def _raise(error):
         )
         raise IOError(msg)
 
-    backoff = 0.01
-    if timeout and timeout / 20 < backoff:
-        backoff = timeout / 20
+    backoff_base = 0.01
+    attempt = 0
 
-    retry_timeout_backoff = random.randrange(140, 160) / 100
-
-    # This starts a thread
     while True:
         try:
-            while deadline - time() > 0:
-
-                async def _():
-                    comm = await connector.connect(
-                        loc, deserialize=deserialize, **connection_args
-                    )
-                    local_info = {
-                        **comm.handshake_info(),
-                        **(handshake_overrides or {}),
-                    }
-                    try:
-                        handshake = await asyncio.wait_for(comm.read(), 1)
-                        write = await asyncio.wait_for(comm.write(local_info), 1)
-                        # This would be better, but connections leak if worker is closed quickly
-                        # write, handshake = await asyncio.gather(comm.write(local_info), comm.read())
-                    except Exception as e:
-                        with suppress(Exception):
-                            await comm.close()
-                        raise CommClosedError() from e
-
-                    comm.remote_info = handshake
-                    comm.remote_info["address"] = comm._peer_addr
-                    comm.local_info = local_info
-                    comm.local_info["address"] = comm._local_addr
-
-                    comm.handshake_options = comm.handshake_configuration(
-                        comm.local_info, comm.remote_info
-                    )
-                    return comm
-
-                with suppress(TimeoutError):
-                    comm = await asyncio.wait_for(
-                        _(), timeout=min(deadline - time(), retry_timeout_backoff)
-                    )
-                    break
-            if not comm:
-                _raise(error)
+            comm = await asyncio.wait_for(
+                connector.connect(loc, deserialize=deserialize, **connection_args),
+                timeout=deadline - time(),
```

Comment:
This drops the change added in https://github.com/dask/distributed/pull/3104 where the initial connect timeout was set lower than the full timeout to kill connect attempts that hung due to dynamic DNS race conditions. In #4167 I made the initial attempt longer (the existing progression was too short for my test case, no attempt was long enough to succeed). I think we still want our *connect* timeouts to be shorter than the full deadline (to allow a few connect attempts), and that we should only retry on connects not on handshakes here (once data is able to send over the wire, we should abort retrying to the application code above this). Perhaps the connect timeouts should start as a fraction of the full timeout (with a min of a couple seconds), and backoff quickly to the remaining timeout?

Reply:
> This drops the change added in #3104 where the initial connect timeout was set lower than the full timeout to kill connect attempts that hung due to dynamic DNS race conditions

Sure, wasn't aware of what exactly was going on here. I can cap these intermediates with a fraction of the total

> and that we should only retry on connects not on handshakes here (once data is able to send over the wire, we should abort retrying to the application code above this)

I disagree on this point. The handshake is a rather internal detail of the connect. How is the application supposed to deal with a failed handshake, let alone detect it?

Reply:
> How is the application supposed to deal with a failed handshake, let alone detect it?

It's not a firm stance. The retries here on connect were to deal with connection-level issues. If a socket connects, but the handshake fails, I'd argue that retrying here isn't worth it (we had a valid connection, then we lost it), and the retry should happen higher up at the application level. Not a big deal though, adding a retry on handshake failure adds no harm, I just don't think it's necessary.

Reply:
understood. I think this is true for actual failures but with the code on master, there is also the possibility that these handshakes simply timeout. At the very least we should increase the timeouts of these handshakes since the 1s might just not be enough if the system is under stress.

I figured I'd just deal with all at once and not distinguish between the cases for the sake of reduced complexity.

Reply:
>  At the very least we should increase the timeouts of these handshakes since the 1s might just not be enough if the system is under stress.

I agree with this, and agree with including them with the larger timeout. I'm just arguing that in the case that we get a successful connection but fail to handshake, we shouldn't retry here. Something like (pseudocode):

```python
while time_remaining_until_deadline:
    try:
        comm = try_to_connect_with_an_increasing_connect_timeout()
        break
    except TimeoutError:
        continue

handle_handshake_in_remaining_time_until_deadline_or_timeout(comm)
```

If the connection succeeds, but then fails during handshake, that indicates either:

- The handshake will always fail, some incompatibility between nodes
- The connection was valid, but then was dropped. This could happen after the comm is returned just as easily, I don't feel like we should handle that here.
- The handshake timed out, which means we've reached the full deadline

None of these IMO should be retried here inside the comm code.

That said, if it's easier to express the timeout logic this way that's fine with me, but I don't think we should expend effort to loop on handshake failure here.

Reply:
Ok, finally understood this. I think removing the handshakes from the retries is ok-ish. There may be other comm errors, of course but they can probably be handled further up the stack.
with the `deadline - time()` logic, removing them from retries does not really make anything more complex imo. I changed the code accordingly